### ISSUE TITLE
Add ML Kit support

### ISSUE THREAD

**Issue by EddyVerbruggen** [EddyVerbruggen/nativescript-plugin-firebase#699]
Yesterday at Google I/O a new feature was announced for Firebase: [ML Kit](https://firebase.google.com/docs/ml-kit/). Let's see how neatly we can integrate its features into this plugin.

## Feature checklist
_If you're curious about the implementation: I'm continuously pushing my changes to [this branch](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/tree/issue699-mlkit-support)._

_The checkboxes indicate my progress (on both platforms). Some items may move to different issues/versions._

### Generic
- [x] Bump the native Firebase SDK's to the versions that support ML Kit
- [x] Fix the Android build to support different sub-SDK versions (as that's a new thing for Firebase)
- [x] Fix the iOS build to correctly load the TensorFlowLite SDK
- [x] Fix the iOS build to correctly load the required bundles for offline ML support (likely a {N} issue)
- [x] Allow configuration of frame-processing-throttling (less strain on the device)
- [x] Add ML Kit installation prompts to the installer (we don't want to add it if folks don't want it)
- [x] Demo's
- [x] Documentation
- [x] Regression-test all non-ML features (because of the SDK bumps)

### [Text Recognition](https://firebase.google.com/docs/ml-kit/recognize-text)
- [x] Recognize text from an image on the device
- [x] Recognize text from a camera stream (🤙[realtime OCR, anyone?](https://twitter.com/eddyverbruggen/status/994551241182273538))
- [x] Expose position information for each detected text block
- [x] Recognize text from an image in the cloud, for better accuracy (but less speed)

### [Barcode Scanning](https://firebase.google.com/docs/ml-kit/read-barcodes)
- [x] Recognize barcodes from an image on the device (there's no cloud option)
- [x] Not only return the encoded _value_, but also return the _type_ of barcode to the client
- [x] (Continuously) recognize barcodes from a camera stream
- [x] Limit barcode formats to certain types (speeds up recognition)
- [x] Create an [embedded barcodescanner](https://twitter.com/eddyverbruggen/status/994531713287118848) (😱be very afraid, barcodescanner plugin!)

### [Face Detection](https://firebase.google.com/docs/ml-kit/detect-faces) 👩‍👧‍👦
- [x] (Continuously) detect faces from a camera stream (only create a photo when [everyone in the frame has their eyes open and smiles](https://twitter.com/eddyverbruggen/status/994627242524332032) 😀?)
- [x] Detect faces from an image on the device (there's no cloud option) - _can't get this to work on iOS currently, will mark it as a known issue for now_

### [Image Labeling](https://firebase.google.com/docs/ml-kit/label-images) 🏷 🏷
- [x] Get labels from an image on the device
- [x] (Continuously) label images from a camera stream
- [x] Get labels from an image in the cloud, for better accuracy (but less speed)

### [Landmark Recognition](https://firebase.google.com/docs/ml-kit/recognize-landmarks) 🗼🗽
- [x] Recognize well-known landmarks in an image in the cloud (there's no device option)

### [Custom Models](https://firebase.google.com/docs/ml-kit/use-custom-models) (moved to 6.1.0, see #702)
~Use a custom TensorFlow Lite model, on the device~
~Use a custom TensorFlow Lite model, stored in Firebase (so users always have the latest model)~

**Comment by jlooper** [comment]
Yasssss!!!! So excited by this!

**Comment by psrcek** [comment]
Is there some sort of beta on npm yet? I'd really like to use this in my app, or at least try it out. If there's no beta, can you reference some build instructions please, i wasnt able to find any :stuck_out_tongue_closed_eyes:

**Comment by EddyVerbruggen** [comment]
@psrcek Thanks for asking! As you found out you can create a local package of this branch, but you're best off waiting a day or two because I'm currently hammering down on this feature like a madman. Expect it to be more stable very soon!

**Comment by psrcek** [comment]
So what you're saying is that you will make a release after a day or two with mlkit included?

**Comment by EddyVerbruggen** [comment]
My comment was about trying [this branch](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/tree/issue699-mlkit-support).

I'll make a release once I'm happy with all parts of the release and there's no regression. That's not gonna happen in two days, but I hope to find enough time to release it next week.

**Comment by EddyVerbruggen** [comment]
Ha, @psrcek very cool that you've already tried it! Indeed, that feature is not exposed through the plugin yet but it can easily be added. I'll add it to the featurelist above.

To anyone interested, I just published a beta release: `tns plugin add nativescript-plugin-firebase@next`. It contains most features I planned in the original post above. Note that there's no documentation yet, but there's a [very elaborate demo app](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/tree/issue699-mlkit-support/demo-ng) that should help you get started.

@psrcek In your case you seem to know how to install the plugin from a cloned Git repo so don't bother installing the beta.

**Comment by EddyVerbruggen** [comment]
@psrcek The [ML Kit branch](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/tree/issue699-mlkit-support) has been updated to give you more details when recognizing text locally on a device.

If you're using a still image then it's good to know the function name has changed from `recognizeText` to `recognizeTextLocal` because there's now also `recognizeTextCloud`. The Cloud bits of the FIrebase SDK are in 'preview' mode and will probably change. Text recognition is probably more accurate but because of the unstable API in their SDK I simply return the raw text back to the client (for now).

**Comment by jlooper** [comment]
Over-the-moon excited to try this, just in time for My Epic Next Demo 💃 #ThanksEddy

**Comment by psrcek** [comment]
@EddyVerbruggen I used the new text recognition and I have a suggestion: your interfaces should look like this IMO to reduce confusion, somewhat match the java implementation and make lines accessible (they aren't right now):

```typescript
export interface MLKitRecognizeTextResultBounds {
	x: number;
	y: number;
	width: number;
	height: number;
}

export interface MLKitRecognizeTextResultElement {
	text: string;
	bounds: MLKitRecognizeTextResultBounds;
}

export interface MLKitRecognizeTextResultLine {
	text: string;
	bounds: MLKitRecognizeTextResultBounds;
	elements: MLKitRecognizeTextResultElement[];
}

export interface MLKitRecognizeTextResultBlock {
	text: string;
	bounds: MLKitRecognizeTextResultBounds;
	lines: MLKitRecognizeTextResultLine[];
}

export interface MLKitRecognizeTextOnDeviceResult extends MLKitResult {
	blocks: MLKitRecognizeTextResultBlock[];
}
```

**Comment by psrcek** [comment]
I made a [pull request](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/pull/708) with changes to the android code, I didn't change the iOS code though as I don't have a mac so I haven't even tried making anything with iOS and I don't know how to code for iOS. :rofl:

**Comment by EddyVerbruggen** [comment]
Hey @psrcek I'll check it out.. I was actually implementing your changes myself and have the iOS ready so should be easy :) Thanks!

**Comment by EddyVerbruggen** [comment]
**6.0.1 has been RELEASED!**

- **iOS**: Please run `pod repo update` from the command line.
- **Android**: Please update you Play services (Android Studio > Tools > SDK manager).
- **Both platforms**: Make sure to remove your 'platforms' folder so the old native SDK's are removed.

To configure and play with ML Kit, follow the instructions in [the doc](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/blob/master/docs/ML_KIT.md).

In case of issues: please post them here, on GitHub. Not on Slack/Twitter/Email. Thanks!

**Comment by psrcek** [comment]
@EddyVerbruggen I made my friend lend me his mac and tried to compile my project on iOS, but I keep getting this message spammed in console whenever i try to use any live preview components:
```
An empty result returned from from GMVDetector for VisionTextDetector.
```
The preview is displayed just fine, but does not detect anything.
(keep in mind that my project works fine on android)

**Comment by EddyVerbruggen** [comment]
@psrcek Is that with the demo app mentioned at the top of the ML Kit readme in this repo (linked in my previous post)? If not, please try that first. The steps are in the readme. It's working fine on my iPhone X...

Otherwise please share a repo reproducing this issue. For instance, do you have [all these set to `true`](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/blob/02cb7d3db4311618aa09c9fb1972528e97bb69fa/demo-ng/firebase.nativescript.json#L15-L20)?

Btw please open a new issue as this one is closed and shipped ;)

**Comment by psrcek** [comment]
Yeah, oops i only had "ml_kit" set to true, you must have added those at some point in between me trying this :stuck_out_tongue:

**Comment by EddyVerbruggen** [comment]
That's a very good point, I did add those (I don't want to ship stuff with your app you don't actually need). 👍

**Comment by coderaven** [comment]
I just want to leave a comment here that you all sirs are heroes to us! Thanks for making this plugin happen! 😄
